Endpoints

HeDo88TH opened the pull request:
Endpoints are now almost complete: MVP reached.

HeDo88TH commented:
I'm fixing the final issues with merge and I'm updating the docs to better reflect the last changes

HeDo88TH commented:
Checks are failing because we should install ddb in CI pipeline before running tests

pierotofy commented on Registry.Web/Services/Adapters/ObjectsManager.cs:
```diff
@@ -0,0 +1,244 @@
+            if (objInfo == null)
+                throw new NotFoundException($"Cannot find '{path}' in storage provider");
+            
+            await using var memory = new MemoryStream();
```
What if the file is really big?

HeDo88TH replied:
Both #33 and #31 are work in progress

pierotofy commented on Registry.Ports/DroneDB/Models/DdbObject.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace Registry.Ports.DroneDB.Models
+{
+    public class DdbObject
```
This should be renamed `DdbEntry`.

HeDo88TH replied:
Fixed in cf6a65a

pierotofy commented:
Making the following request:

`POST http://localhost:5000/ddb/share/init` without auth info returns an empty response (error code is correct, 401)

It should return:

```
{
    "error": "Unauthorized"
}
```

pierotofy commented:
`POST /ddb/share/init` should be changed a bit:
 - No JSON fields. Should use form data (just like `/users/authenticate`).
 - The only fields it should expose are: [`tag` (optional), `password` (optional)]
 - It's the task of the Registry to parse `tag` (which is a representation of `organization/dataset`). 
 - If no `tag` is specified, the dataset should be placed in the system/public organization and a random dataset UUID should be picked.
 - A password field needs to be added to the dataset model.
 - A user should not be able to specify `slug` manually. These are always computed based on the `name`.

`ddb share [FILES|DIRECTORIES] [FLAGS] [--tag orgname/dataset] [--password <pwd>]`

pierotofy commented:
This will probably get fixed with the changes proposed above, but for the sake of testing, calling `localhost:5000/ddb/share/init` with:

```
{
    "organization": {
        "id": "test",
        "name": "Test organization",
        "description": "This is a test organization",
        "isPublic": true
    },
    "dataset": {
        "slug": "first",
        "name": "First dataset",
        "description": "This is my first dataset",
        "isPublic": true
    }
}
```

Returns:

```
{
    "error": "Value cannot be null. (Parameter 'source')"
}
```

pierotofy commented:
We need to be **really** careful about using `POST /ddb/share/upload/{token}/{path/to/file}`. {path/to/file} can contain unicode characters, which would force the client to URL-encode the path.

Perhaps `path` should be just a form data field?

https://stackoverflow.com/questions/2742852/unicode-characters-in-urls

pierotofy commented:
`POST localhost:5000/ddb/` should accept form data, not JSON.

pierotofy commented:
`DELETE http://localhost:5000/ddb/test2` should return something like:

```
{
    "success": true
}
```

Or some other confirmation.

pierotofy commented:
`POST localhost:5000/ddb/test2/ds/aukerman` or even `POST localhost:5000/ddb/test2/ds` return `405 not allowed` or `404 not found`. Note I'm not sure this endpoint will be necessary in the end. Meta endpoints like `ddb/share` might be the only ones that allow the creation of a dataset.

pierotofy commented:
`PUT http://localhost:5000/ddb/test2` return no content (should return something affermative like `{ "success": true }' or a copy of the newly modified model?

pierotofy commented:
If I call `POST http://localhost:5000/ddb/` with:

```
{
    "id": 3,
    "name": "Test organization",
    "creationDate": 0
}
```

It explodes. I think a person should be able to create an organization by simply specifying the ID and name (optionally), all other fields should be prepopulated.

pierotofy commented:
I think these are good starters! Let me know if there are questions. :+1:

HeDo88TH commented on Registry.Ports/DroneDB/Models/DdbInfo.cs:
```diff
@@ -0,0 +1,53 @@
+        public DdbInfoType Type { get; set; }
+    }
+
+    public enum DdbInfoType
```
Fixed in ab2495b

HeDo88TH commented on Registry.Ports/DroneDB/Models/DdbObject.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace Registry.Ports.DroneDB.Models
+{
+    public class DdbObject
```
Yes, one for concern.

HeDo88TH commented on Registry.Web.Test/DdbFactoryTest.cs:
```diff
@@ -0,0 +1,246 @@
+        private const string TestDataFolder = @"Data/Ddb";
+        private const string DdbTestDataFolder = @"Data/DdbTest";
+
+        private const string Test1ArchiveUrl = "https://digipa.it/wp-content/uploads/2020/08/Test1.zip";
```
Changed in 432d871

HeDo88TH commented on Registry.Web.Test/ObjectManagerTest.cs:
```diff
@@ -0,0 +1,291 @@
+
+        private const string BaseTestFolder = "ObjectManagerTest";
+
+        private const string Test1ArchiveUrl = "https://digipa.it/wp-content/uploads/2020/08/Test1.zip";
```
Fixed in 432d871

HeDo88TH commented:
Switched from JSON to form-data as input parameters format for all web services (24ca60a)

HeDo88TH commented on Registry.Web/Controllers/ObjectsController.cs:
```diff
@@ -0,0 +1,127 @@
+                if (file == null)
+                    return BadRequest(new ErrorResponse("No file uploaded"));
+
+                await using var memory = new MemoryStream();
```
Both #33 and #31 are work in progress

HeDo88TH commented on Registry.Web/Controllers/ShareController.cs:
```diff
@@ -0,0 +1,99 @@
+                if (file == null)
+                    return BadRequest(new ErrorResponse("No file uploaded"));
+                
+                await using var memory = new MemoryStream();
```
Both #33 and #31 are work in progress

HeDo88TH commented on Registry.Web/Controllers/ObjectsController.cs:
```diff
@@ -0,0 +1,127 @@
+
+    [Authorize]
+    [ApiController]
+    [Route("ddb/{orgId:alpha}/ds/{dsId:alpha}")]
```
Fixed in a3140a1

HeDo88TH commented:
> `DELETE http://localhost:5000/ddb/test2` should return something like:
> 
> ```
> {
>     "success": true
> }
> ```
> 
> Or some other confirmation, instead of an empty response? I know we can look at the status codes, but it might be better to send some content along with the status code.

As stated in https://stackoverflow.com/questions/2342579/http-status-code-for-update-and-delete the correct status code is 204 No Content, that has... no content.

pierotofy commented:
Mm, maybe we can return `200` instead then?

HeDo88TH commented:
Let's stick to the standard. If a DELETE fails It will return 4XX with error message, otherwise everything went well.

HeDo88TH commented on Registry.Adapters.DroneDB/Ddb.cs:
```diff
@@ -0,0 +1,90 @@
+            var res = RunCommand($"{InitCommand} -d \"{tempPath}\"");
+
+            // This way we preserve the ACL
+            File.WriteAllBytes(path, File.ReadAllBytes(Path.Combine(tempPath, ".ddb", "dbase.sqlite")));
```
Fixed in bbe15fe

HeDo88TH commented:
> This will probably get fixed with the changes proposed above, but for the sake of testing, calling `localhost:5000/ddb/share/init` with:
> 
> ```
> {
>     "organization": {
>         "id": "test",
>         "name": "Test organization",
>         "description": "This is a test organization",
>         "isPublic": true
>     },
>     "dataset": {
>         "slug": "first",
>         "name": "First dataset",
>         "description": "This is my first dataset",
>         "isPublic": true
>     }
> }
> ```
> 
> Returns:
> 
> ```
> {
>     "error": "Value cannot be null. (Parameter 'source')"
> }
> ```

Modified in 08d102d

HeDo88TH commented:
All the issues and proposals have been addressed. I'm going to write some end-to-end tests to ensure that the overall behaviour is ok, then we can merge it @pierotofy

pierotofy commented:
Excellent! I will be testing the new changes after the tests are pushed.

pierotofy commented:
`POST /ddb/share/init` with:

`tag=test/test` (where test is a valid org):

Returns:

```
{
    "error": "Invalid dataset"
}
```

Shouldn't I be allowed to initiate a share upload to the admin organization?

pierotofy commented:
`POST localhost:5000/ddb/share/upload/CgF0IsVVIWkXEDEKFiHOUmg9FQiFYVe6/DJI.JPG`

With a single image (JPG).

Returns:

```
{
    "type": "https://tools.ietf.org/html/rfc7231#section-6.5.13",
    "title": "Unsupported Media Type",
    "status": 415,
    "traceId": "|6452c39b-49aed024ec18ad75."
}
```

If I choose `binary` from Postman. I think this is wrong (docs not clear).

This should be a multipart/form-data file upload (what's the key for the file?)

Anyway, if I try form-data upload I get:

```
{
    "error": "Sqlite database not found"
}
```

I also think we should place the `path` in the form-data body instead of the URL.

pierotofy commented:
Let's not add more things to this PR, but for the very near future: https://github.com/DroneDB/Registry/issues/38

HeDo88TH commented:
I rolled back b360492 and 25f1ca8 because there is no need to return an object when the HTTP code is already stating it.
I fixed 5aac056, some tests were broken and some logic too.
I added back the fields DatasetName and DatasetDescription to ShareInitDto as optional if the clients wants to specify more details about the dataset.
I added a new TagDto class that handles the parsing of tags like "organization/dataset" and I reorganized some utils methods in extenders methods.

pierotofy commented:
Maybe it's just my environment, but I still cannot get a file upload going:

`POST /ddb/share/init`

Passing: `tag=test2/test2` as form-data. The org `test2` exists.

```
      Dataset 'test2' not found, creating it
fail: Registry.Web.Controllers.ShareController[0]
      Exception in Share controller Init('Registry.Web.Models.DTO.ShareInitDto')
Registry.Web.Exceptions.NotFoundException: Cannot find dataset
   at Registry.Web.Services.Adapters.WebUtils.GetDatasetAndCheck(String orgSlug, String dsSlug, Boolean retNullIfNotFound) in D:\RegistryApp\Registry.Web\Services\Adapters\WebUtils.cs:line 81
   at Registry.Web.Services.Adapters.ShareManager.Initialize(ShareInitDto parameters) in D:\RegistryApp\Registry.Web\Services\Adapters\ShareManager.cs:line 123
   at Registry.Web.Controllers.ShareController.Init(ShareInitDto parameters) in D:\RegistryApp\Registry.Web\Controllers\ShareController.cs:line 35
```

pierotofy commented:
This is still not working. :(

`POST /ddb/share/upload/<validToken>/IMG_20160904_130804.jpg`

```
{
"error": "Invalid path"
}
```

Side note, path is still encoded in the URL. I thought we addressed this in a previous comment during this review.

>> I also think we should place the path in the form-data body instead of the URL.

pierotofy reviewed (changes requested):
1) Make sure nominal test paths work
2) Make sure `path` is encoded in the form-data body, not URL

HeDo88TH commented:
> 1. Make sure nominal test paths work

I tested with postman and It works on my machine. I added logging to a text file to improve issue tracking (d013f57), when something breaks attach the log file to better identify the issue.

> 2. Make sure `path` is encoded in the form-data body, not URL

Addressed in d2d8fa7, I will update the docs as soon as possible.

pierotofy commented:
:+1: 

Excellent!

I've made a mistake while testing; (was using the wrong IP address in postman).
